pkg/agent/controller/traceflow/packetin.go

```diff
@@ -237,7 +237,13 @@ func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (*crdv1alpha1.Tracefl
 	if tableID == openflow.L2ForwardingOutTable.GetID() {
 		ob := new(crdv1alpha1.Observation)
 		tunnelDstIP := ""
-		isIPv6 := c.nodeConfig.NodeIPv6Addr != nil
+		var isIPv6 bool
+		// On dual stack, decide according to packet.
+		if c.nodeConfig.NodeIPv4Addr != nil && c.nodeConfig.NodeIPv6Addr != nil {
+			isIPv6 = etherData.Ethertype == protocol.IPv6_MSG
+		} else {
+			isIPv6 = c.nodeConfig.NodeIPv6Addr != nil
+		}
 		if match := getMatchTunnelDstField(matchers, isIPv6); match != nil {
```

Comment:
Maybe we'd better get both IPv4 tunnel dst and IPv6 tunnel dst, and set `tunnelDstIP` with valid one.

You can also add verification in `traceflow_test.go:testTraceflowInterNode`, to check if the traceflow observation has expected tunnelDstIP.

Reply:
Hi,
Even if we get both IPv4 tunnel dst and IPv6 tunnel dst, `tunnelDstIP` value will be decided according to `isIPv6`. The current code sets the value of `tunnelDstIP` according to `isIPv6` directly without getting both tunnel dst. So, why to get both tunnel dst explicitly?

`testTraceflowInterNode` currently runs in both encap mode and no encap mode. If tunnelDstIP is added then we have to make the test run only in encap mode. Is it ok to change this test to run only in encap mode?

Reply:
whether it is dual stack / non dual stack, in both the cases handling can be done with isIPv6 flag with right packet type, With this logic isIPv6 flag remain false if packet is not of ipv6 type irrespective of stack and hence tunnelDstIP will be assigned with ipv4 and vice versa.

UT testing can be extended to see the right values in tunnelDstIP, if the mode is encap.

Reply:
> Hi, Even if we get both IPv4 tunnel dst and IPv6 tunnel dst, `tunnelDstIP` value will be decided according to `isIPv6`. The current code sets the value of `tunnelDstIP` according to `isIPv6` directly without getting both tunnel dst. So, why to get both tunnel dst explicitly?
> 
> `testTraceflowInterNode` currently runs in both encap mode and no encap mode. If tunnelDstIP is added then we have to make the test run only in encap mode. Is it ok to change this test to run only in encap mode?

Because reading both ipv4/ipv6 tunnel dst is more simple to decide the tunnel dst, and it won't depend on the flow logic in OVS table.
Currently only IPv4 tunnel dst is used in DualStack env. Please check: https://github.com/antrea-io/antrea/blob/main/pkg/agent/openflow/client.go#L485

You can add the TunnelDstIP into `testTraceflowInterNode` like
```
node2TunnelDstIP = ""
switch encapMode {
case config.TrafficEncapModeEncap:
	node2TunnelDstIP = nodeIP(nodeIdx1)
case config.TrafficEncapModeHybrid:
	// check more
}
...
						{
							Component:     v1alpha1.ComponentForwarding,
							ComponentInfo: "Output",
							Action:        v1alpha1.ActionForwarded,
							TunnelDstIP:   node2TunnelDstIP,
						},
```
And add the comparison into `compareObservations`

Reply:
@wenyingd 
Do we use peer Node IPv4 as tunnelDst for IPv4 packet and peer Node IPv6 as tunnelDst for IPv6 packet?

Reply:
We prefer to use tunnel format IPv4OverIPv4 and IPv6OverIPv6.